FirebaseDatabase/Tests/Integration/FIRDatabaseQueryTests.m

```diff
@@ -4157,6 +4157,310 @@ - (void)testGetForNodeReturnsCorrectValue {
+    [ref getDataWithCompletionBlock:^(NSError* _Nullable error,
+                                      FIRDataSnapshot* _Nullable snapshot) {
+      XCTAssertNil(error);
+      XCTAssertEqualObjects(@1, snapshot.value);
+      done = YES;
+    }];
+    WAIT_FOR(done);
+  } @finally {
+    [[ref database] goOnline];
+  }
+}
+
+- (void)testGetResolvesToCacheWhenOnlineAndParentListener {
+  FTupleFirebase* pair = [FTestHelpers getRandomNodePair];
+  FIRDatabaseReference* writer = pair.one;
+  FIRDatabaseReference* reader = pair.two;
+  [self waitForCompletionOf:writer setValue:@{@"a" : @1}];
+  [self snapWaiter:reader
+         withBlock:^(FIRDataSnapshot* snapshot) {
+           XCTAssertEqualObjects(
+               @{@"a" : @1}, snapshot.value);
+         }];
+  __block BOOL done = NO;
+  [[reader child:@"a"]
+      getDataWithCompletionBlock:^(NSError* _Nullable error, FIRDataSnapshot* _Nullable snapshot) {
+        XCTAssertNil(error);
+        XCTAssertEqualObjects(@1, snapshot.value);
+        done = YES;
+      }];
+  WAIT_FOR(done);
+}
+
+- (void)testGetResolvesToCacheWhenOnlineAndSameLevelListener {
+  FTupleFirebase* pair = [FTestHelpers getRandomNodePair];
+  FIRDatabaseReference* writer = pair.one;
+  FIRDatabaseReference* reader = pair.two;
+  [self waitForCompletionOf:writer setValue:@{@"a" : @1}];
+  [self snapWaiter:reader
+         withBlock:^(FIRDataSnapshot* snapshot) {
+           XCTAssertEqualObjects(
+               @{@"a" : @1}, snapshot.value);
+         }];
+  __block BOOL done = NO;
+  [reader
+      getDataWithCompletionBlock:^(NSError* _Nullable error, FIRDataSnapshot* _Nullable snapshot) {
+        XCTAssertNil(error);
+        XCTAssertEqualObjects(
+            @{@"a" : @1}, snapshot.value);
+        done = YES;
+      }];
+  WAIT_FOR(done);
+}
+
+- (void)testGetResolvesToServerCacheWhenListenerIsAvailable {
+  FTupleFirebase* pair = [FTestHelpers getRandomNodePairWithoutPersistence];
+  FIRDatabaseReference* writer = pair.one;
+  FIRDatabaseReference* reader = pair.two;
+  [self waitForCompletionOf:writer setValue:@{@"a" : @1}];
+  __block BOOL done = NO;
+  FIRDatabaseHandle handle = [reader observeEventType:FIRDataEventTypeValue
+                                            withBlock:^(FIRDataSnapshot* _Nonnull snapshot) {
+                                              XCTAssertEqualObjects(
+                                                  @{@"a" : @1}, snapshot.value);
+                                              done = YES;
+                                            }];
+  WAIT_FOR(done);
+  done = NO;
+  @try {
+    [[reader database] goOffline];
+    [reader getDataWithCompletionBlock:^(NSError* _Nullable error,
+                                         FIRDataSnapshot* _Nullable snapshot) {
+      XCTAssertNil(error);
+      XCTAssertEqualObjects(
+          @{@"a" : @1}, snapshot.value);
+      done = YES;
+    }];
+    WAIT_FOR(done);
+    [reader removeObserverWithHandle:handle];
```

Comment:
This seems like the only place where the observer is removed. Do we need to do that in the other tests as well?